src/main/java/com/uber/cadence/internal/compatibility/Thrift2ProtoAdapter.java

```diff
@@ -1293,4 +1294,13 @@ public void SignalWorkflowExecutionWithTimeout(
       throws TException {
     throw new UnsupportedOperationException("not implemented");
   }
+
+  private void handleException(AsyncMethodCallback callback, Exception exception) {
+    if (exception instanceof ExecutionException
```

Comment:
This behavior is different from the non-async methods. They throw `StatusRuntimeException` directly. Could you point me where we return `ExecutionException` wrapped exceptions?

Reply:
Within the async methods we have blocks that look like this:
```
      resultFuture.addListener(
          () -> {
            try {
              com.uber.cadence.api.v1.GetWorkflowExecutionHistoryResponse response =
                  resultFuture.get();
              resultHandler.onComplete(
                  ResponseMapper.getWorkflowExecutionHistoryResponse(response));
            } catch (Exception e) {
              resultHandler.onError(e);
            }
          }
```

If the `resultFuture` [completed with an exception](https://docs.oracle.com/javase/8/docs/api/index.html?java/util/concurrent/Future.html) it will throw an ExecutionException wrapping the actual cause of the failure (the StatusRuntimeException). With our current implementation this has two issues:
- We're completing the resultHandler with the ExecutionException.
- We don't map the StatusRuntimeException to the thrift error type.

So currently if a user should receive an AccessDeniedError they'll receive an ExecutionException containing a StatusRuntimeException.

Reply:
got it. Maybe we can rename it as `handleAsyncException`?

Reply:
Sounds good, done.